modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,541 @@
+            The function to use for the filter. This function should filter the
+            data itself. It accepts either a row or column (depending on the axis argument) and
+            returns True to keep the row/col, and False to drop it.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe filtered by content according to the filter provided by condition.
+        """
+        pass
+
+    @abstractmethod
+    def explode(
+        self,
+        axis: Union[int, Axis],
+        function: Callable,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """
+        Explode data based on the function provided along the specified axis.
+
+        Parameters
+        ----------
+        axis : int or modin.core.dataframe.base.dataframe.Axis
+            The axis to expand over.
+        function : callable
+            The function to use to expand the data. This function should accept one
+            row/column, and return multiple.
+        result_schema : dictionary, optional
+            Mapping from column labels to data types that represents the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the specified axis expanded.
+
+        Notes
+        -----
+        Only one axis can be expanded at a time.
+
+        The user-defined function may increase the number of rows (columns if axis=1),
+        but it should not remove or drop rows.
+        """
+        pass
+
+    @abstractmethod
+    def window(
+        self,
+        axis: Union[int, Axis],
+        reduce_fn: Callable,
+        window_size: int,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """
+        Apply a sliding window operator that acts as a GROUPBY on each window, reducing each window to a single row (column).
+
+        Parameters
+        ----------
+        axis : int or modin.core.dataframe.base.dataframe.Axis
+            The axis to slide over.
+        reduce_fn : callable(rowgroup|colgroup) -> row|col
+            The reduce function to apply over the data.
+        window_size : int
+            The number of row/columns to pass to the function.
+            (The size of the sliding window).
+        result_schema : dictionary, optional
+            Mapping from column labels to data types that represents the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the reduce function applied over windows of the specified
+            axis.
+
+        Notes
+        -----
+        The user-defined reduce function must reduce each window’s column
+        (row if axis=1) down to a single value.
+        """
+        pass
+
+    @abstractmethod
+    def groupby(
+        self,
+        axis: Union[int, Axis],
+        by: Union[str, List[str]],
```

Comment:
Should column labels be Hashables, not str?

column_labels are Hashables in mask but str in other methods like this one, infer_types, to_labels, sort_by.

Reply:
Good question - for some reason the doctoring has by as strings specifically. Its a washable in `mask` cause you can `mask` by lambdas etc. @devin-petersohn do we want to keep groupby's `by` argument as a list of column names, or should we make it a hashable to support functions etc?